functions/go/upsert-resource/upsertresource/upsert_resource.go

```diff
@@ -0,0 +1,167 @@
+package upsertresource
+
+import (
+	"strings"
+
+	"sigs.k8s.io/kustomize/kyaml/fn/runtime/runtimeutil"
+	"sigs.k8s.io/kustomize/kyaml/kio/filters"
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// UpsertResource upserts resource to input list of resources
+// resources are uniquely identifies by Group, Kind, Name and Namespace
+type UpsertResource struct {
+	// Resource is the input resource for upsert
+	Resource *yaml.RNode
+}
+
+// Filter implements UpsertResource as a yaml.Filter
+func (ur UpsertResource) Filter(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	replaced, err := ReplaceResource(nodes, ur.Resource)
+	if err != nil {
+		return nodes, err
+	}
+	if !replaced {
+		return AddResource(nodes, ur.Resource)
+	}
+	return nodes, nil
+}
+
+// ReplaceResource checks if the inputResource matches with any of the resources in nodes list
+// and replaces it with inputResource
+// nodes are matched based on Group, Kind, Name and Namespace
+func ReplaceResource(nodes []*yaml.RNode, inputResource *yaml.RNode) (bool, error) {
+	found := false
+	inputMeta, err := inputResource.GetMeta()
+	if err != nil {
+		return false, err
+	}
+	for i := range nodes {
+		rMeta, err := nodes[i].GetMeta()
+		if err != nil {
+			return false, err
+		}
+		// skip processing resource if it is a local config
+		if IsLocalConfig(rMeta) {
+			continue
+		}
+		// check if there is a match and replace the resource
+		if IsSameResource(inputMeta, rMeta) {
+			nodes[i], err = deepCopy(inputResource)
+			if err != nil {
+				return false, err
+			}
+			a := combineInputAndMatchedAnnotations(inputMeta.Annotations, rMeta.Annotations)
+			err = nodes[i].SetAnnotations(a)
+			if err != nil {
+				return false, err
+			}
+			// found a matching resource
+			found = true
```

Comment:
Can we return earlier here?

Reply:
No. Updated the comment.

Reply:
What is the expected behavior there are more than 1 resources with the same identifier (GKNN)?

Reply:
This function replaces all of them as there could be other functions in the pipeline like set-name-prefix which changes the name of the resource after this function.

Reply:
Got it!